### DISCUSSION TITLE
Use wrapper objects instead of mailbox workspaces

### DISCUSSION THREAD

**Discussion by MikeSpreitzer** [kubestellar/kubestellar#595] [Ideas]
The only place where the edge-mc platform _creates_ a runtime variable collection of logical clusters is the mailbox workspaces (it is users of the platform that create inventory management workspaces and workload management workspaces).  This is the one collection that we expect to stress scale significantly.  With non-kcp providers of this collection costing 100s of MB (in both disk and RAM) per logical cluster, we are incented to find an alternative approach.  One alternative would be to collapse all the mailbox workspaces to one cluster and wrap every downsynced and upsynced object with a generic wrapper.  We could have a mailbox namespace for each edge cluster, holding the wrapped objects going to/from that edge cluster.

There are three components that would be affected by this change: the placement translator, the syncer, and the mailboxwatch library.

The placement translator would be easy to adapt.

The mailboxwatch library, in its current form, should be doable without great effort.  There is a potential for trouble here.  The base informer library allows restricting the upstream object collection by label selectors and field selectors; every object supports selectors on the metadata fields for name and namespace.  _If_ the mailboxwatch library were correspondingly configurable, there would be a problem of how to handle those field selectors on namespace and name (since the base object namespace and name will not appear in the wrapped object namespace and name).  Remember that label values are very constrained (see https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set).  However, today the mailboxwatch library does not offer this filtering, so we are currently dodging that bullet.

@yana1205 , how about the syncer?

**Comment by MikeSpreitzer** [comment]
There is also the matter of what wrapper object to use.  We could define one of our own.  We could use OCM's ManifestWork, with exactly one object carried inside each ManifestWork.  That might even enable the use of OCM's klusterlet for downsync.  But OCM does not do upsync, so ...?

**Reply by clubanderson** [reply]
AppWrapper from MCAD might be a better choice since it is aligned with our direction for q3

**Reply by MikeSpreitzer** [reply]
The proposal here is about making a change in an internal mechanism, not a change in the interface that users use.  The latter remains the workload management workspace, into which a user can put whatever objects they want (including AppWrapper and ManifestWork).

**Comment by yana1205** [comment]
As for syncer, the main change is to makes Syncer support unpack/pack the wrapped object similar to what OCM does. 
Does it technically look like the following process?  For downsync part, syncer preprocesses to unpack objects packed as a k8s resource like ManifestWork at a server side. For upsync part, syncer packs upsynced objects (including objects that the status is updated) and sends it to the server.

> With non-kcp providers of this collection costing 100s of MB (in both disk and RAM) per logical cluster, we are incented to find an alternative approach. 

If the provider is kcp, if you know, can I ask how much it costs?

**Comment by MikeSpreitzer** [comment]
@yana1205 : as I learned recently, the OCM klusterlet has a plugin architecture.  We could just use the regular functionality of klusterlet to downsync.  Could we create an upsync plugin?  It would need to be guided by a downsynced SyncerConfig (or something like that) object.

**Reply by yana1205** [reply]
I'm not familiar with how to implement the plugin. I will check it.
